proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,328 @@
+		var br BinaryMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "opCode":
+				br.opCode = v
+				br.opCodes = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+			case "keyPrefix":
+				br.keyPrefix = v
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				panic(fmt.Errorf("Unsupported key: %s", k))
+			}
+		}
+		log.Infof("Parsed BinaryMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, strData []string, offset uint32) (proxylib.OpType, uint32) {
+	log.Infof("OnData with offset %d", offset)
+
+	//TODO: optimise this
+	data := ([]byte(strings.Join(strData, "")))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	headerMissing := headerSize - len(data)
+
+	if headerMissing > 0 {
+		log.Infof("Did not receive full header, need %d more bytes", headerMissing)
+		return proxylib.MORE, uint32(headerMissing)
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
+
+	dataMissing := headerSize + int(bodyLength) - len(data)
+	if dataMissing > 0 {
+		log.Infof("Did not receive full request, need %d more bytes", dataMissing)
+		return proxylib.MORE, uint32(dataMissing)
+	}
+
+	// we don't parse reply traffic for now
+	if reply {
+		if len(data) == 0 {
+			log.Infof("ignoring zero length reply call to onData")
+			return proxylib.NOP, 0
+		}
+
+		log.Infof("reply, passing %d bytes", uint32(len(data)))
+		return proxylib.PASS, uint32(len(data))
+	}
+
+	opcode, key, err := p.parse(data)
+	if err != 0 {
+		return proxylib.ERROR, uint32(err)
+	}
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "binarymemcached",
+			Fields: map[string]string{
+				"opcode": strconv.Itoa(int(opcode)),
+				"key":    key,
+			},
+		},
+	}
+
+	matches := p.connection.Matches(data)
```

Comment:
Here you could pass pre-parsed data (e.g., `{ opcode, key }`) instead so that the incoming message would only parsed once. Now it is effectively parsed again when matching against each rule.

Reply:
done